include/graphblas/reference/blas3.hpp

```diff
@@ -916,6 +916,206 @@ namespace grb {
+		class Operator,
+		typename InputType1, typename InputType2, typename OutputType,
+		typename Coords,
+		typename RIT, typename CIT, typename NIT
+	>
+	RC maskedOuter(
+		Matrix< OutputType, reference, RIT, CIT, NIT > &A,
+		const Matrix< OutputType, reference, RIT, CIT, NIT > &mask,
+		const Vector< InputType1, reference, Coords > &u,
+		const Vector< InputType2, reference, Coords > &v,
+		const Operator &mul = Operator(),
+		const Phase &phase = EXECUTE,
+		const typename std::enable_if<
+			grb::is_operator< Operator >::value &&
+			!grb::is_object< InputType1 >::value &&
+			!grb::is_object< InputType2 >::value &&
+			!grb::is_object< OutputType >::value,
+			void >::type * const = nullptr
+	) {
+		// static checks
+		NO_CAST_ASSERT( ( !(descr & descriptors::no_casting) ||
+				std::is_same< typename Operator::D1, InputType1 >::value
+			), "grb::maskedOuter",
+			"called with a prefactor vector that does not match the first domain "
+			"of the given multiplication operator" );
+		NO_CAST_ASSERT( ( !(descr & descriptors::no_casting) ||
+				std::is_same< typename Operator::D2, InputType2 >::value
+			), "grb::maskedOuter",
+			"called with a postfactor vector that does not match the first domain "
+			"of the given multiplication operator" );
+		NO_CAST_ASSERT( ( !(descr & descriptors::no_casting) ||
+				std::is_same< typename Operator::D3, OutputType >::value
+			), "grb::maskedOuter",
+			"called with an output matrix that does not match the output domain of "
+			"the given multiplication operator" );
+		static_assert( ( !(descr & descriptors::invert_mask)
+			),
+			"grb::maskedOuter: invert_mask descriptor cannot be used ");
+#ifdef _DEBUG
+		std::cout << "In grb::maskedOuter (reference)\n";
+#endif
+
+		const size_t nrows = size( u );
+		const size_t ncols = size( v );
+
+		const size_t m = grb::nrows( mask );
+		const size_t n = grb::ncols( mask );
+
+		assert( phase != TRY );
+		if( nrows != grb::nrows( A ) || nrows != m ) {
+			return MISMATCH;
+		}
+
+		if( ncols != grb::ncols( A ) || ncols != n ) {
+			return MISMATCH;
+		}
+
+
+		const auto &mask_raw = internal::getCRS( mask );
+
+		size_t nzc = 0;
+		for( size_t i = 0; i < nrows; ++i ) {
+			if( internal::getCoordinates( u ).assigned( i ) ) {
+				for( auto k = mask_raw.col_start[ i ]; k < mask_raw.col_start[ i + 1 ]; ++k ) {
+					const size_t k_col = mask_raw.row_index[ k ];
+					if( internal::getCoordinates( v ).assigned( k_col ) ) {
+						nzc++;
+					}
+				}
+			}
+		}
+
+		if( phase == RESIZE ) {
+			return resize( A, nzc );
+		}
+
+		assert( phase == EXECUTE );
+		if( capacity( A ) < nzc ) {
+#ifdef _DEBUG
+			std::cout << "\t insufficient capacity to complete "
+				"requested masked outer-product computation\n";
+#endif
+			const RC clear_rc = clear( A );
+			if( clear_rc != SUCCESS ) {
+				return PANIC;
+			} else {
+				return FAILED;
+			}
+		}
+
+		grb::Monoid<
+			grb::operators::left_assign< OutputType >,
+			grb::identities::zero
+		> monoid;
```

Comment:
Shouldn't it be a template parameter instead ?

Reply:
This monoid can be completely removed, thanks.